### ISSUE TITLE
Bignum: Prevent mpi_mod_write from corrupting the input

### ISSUE THREAD

**Issue by yanesca** [Mbed-TLS/mbedtls#6679]
### Suggested enhancement

Change `mbedtls_mpi_mod_raw_to/from_mont_rep()` to take the in and out parameters separately, instead of a single in/out parameter. This would allow mpi_mod_write to do the conversion in a temporary location.

### Justification

Error in the conversion could make the write function corrupt its input parameter. Also, converting parameters in place for writing them out would prevent writing out values residing in static memory.

**Comment by gilles-peskine-arm** [comment]
`mbedtls_mpi_mod_write` is problematic because it writes to a parameter that is morally const. Its input is a `const mbedtls_mpi_mod_residue *`, but it modifies the limb array. This is valid C, but contradicts the general principle that we don't modify parameters conveyed through const objects. The code only gets away with it because C doesn't let us express “this is const and so is what it's pointing to”.

Another piece of technical debt in `mbedtls_mpi_mod_write` is that it should convert from whatever the representation of the modulus is, calling `mbedtls_mpi_mod_raw_modulus_to_canonical_rep`, instead of assuming that whatever isn't `MBEDTLS_MPI_MOD_REP_MONTGOMERY` is canonical.

**Comment by yanesca** [comment]
@gilles-peskine-arm that is a good point. Does this mean that you are happy with changing `mpi_mod_write()` to do the conversion in a temporary location? (This would also mean that we can write out the value from the temporary and there is no need to convert back.)

**Comment by gilles-peskine-arm** [comment]
> Does this mean that you are happy with changing mpi_mod_write() to do the conversion in a temporary location?

Yes. Either that or it allocates internally. I don't see another solution. It's not very pretty, but either is better than modifying a data structure that looks like it's read-only.

**Comment by yanesca** [comment]
(By "do the conversion in a temporary location" I meant allocating it internally.)

**Comment by yanesca** [comment]
Refactoring the conversions is being done here:
https://github.com/Mbed-TLS/mbedtls/issues/6815
